File: python-analise-dados/python/funcao-compara-rna.py
# Comparando RNAs Humano e Bactéria 
import os
# lista para armazenar as letras dos nucleotídeos
nucleotideos=['A', 'T', 'C', 'G']

def carregar_sequencia(caminho_fasta):
    """Carrega sequência do arquivo FASTA, removendo cabeçalho e quebras de linha."""
    with open(caminho_fasta) as arquivo:
        linhas = arquivo.readlines()
    
    # linha.strip() remove espaços em branco nas extremidades
    sequencia = "".join(linha.strip() 
                        
    # Remove cabeçalho (linhas que começam com '>') e junta o restante                    
    for linha in linhas
      if not linha.startswith(">"))
    
    return sequencia.upper()


def criar_dicionario_dinucleotideos():
    """Cria dicionário com todos os dinucleotídeos possíveis inicializados com 0."""
    dicDinucleotideos = {}
    for i in nucleotideos:
        for j in nucleotideos:
            dicDinucleotideos[i + j] = 0
    return dicDinucleotideos


def contar_dinucleotideos(sequencia, dicionario):
    """Conta a frequência de dinucleotídeos na sequência."""
    for k in range(len(sequencia) - 1):
        par = sequencia[k] + sequencia[k + 1]
        if par in dicionario:
            dicionario[par] += 1
    return dicionario




def gerar_html(contagem, arquivo_saida="bacteria.html"):
    """Gera arquivo HTML com gráfico visual dos dinucleotídeos."""
    # Obtém o diretório do script atual
    diretorio_script = os.path.dirname(os.path.abspath(__file__))
    caminho_html = os.path.join(diretorio_script, arquivo_saida)
    
    with open(caminho_html, "w") as saida:
        saida.write("<html><body>\n")
        i = 1
        max_valor = max(contagem.values())
        
        for dinucleotideo, valor in contagem.items():
            # se max_valor for maior que 0, calcula transparência proporcional
            # se max_valor nao for maior que 0, transparência é = 0
            transparencia = valor / max_valor if max_valor > 0 else 0
            
            # cria a div com a cor azul e transparência proporcional
            # essa div usa a transparencia no canal alpha do RGBA
            saida.write(
                f"<div style='width:100px; border:1px solid #111; height:100px; float:left;"
                f" background-color:rgba(0, 0, 255, {transparencia})'>"
                f"{dinucleotideo}-{valor} </div>\n"
            )
            
            # A cada 4 divs, cria uma nova linha
            # i = contador de divs
            if i % 4 == 0:
                saida.write("<div style='clear:both'></div>\n")
            # incrementa o contador
            i += 1
        
        saida.write("</body></html>")
    
    print(f"✅ HTML salvo em: {caminho_html}")


# Função principal que orquestra toda a análise de RNA.
def analisar_rna(caminho_fasta, arquivo_saida="bacteria.html"):
    """Função principal que orquestra toda a análise de RNA."""
    print(f"Carregando sequência de {caminho_fasta}...")
    sequencia = carregar_sequencia(caminho_fasta)
    
    print("Criando dicionário de dinucleotídeos...")
    dic = criar_dicionario_dinucleotideos()
    print(f"Dicionário inicial: {dic}\n")
    
    print("Contando dinucleotídeos...")
    dic = contar_dinucleotideos(sequencia, dic)
    print(f"Contagem final: {dic}\n")
    
    print(f"Gerando arquivo HTML: {arquivo_saida}...")
    gerar_html(dic, arquivo_saida)
    print("✅ Análise concluída com sucesso!")


# Execução principal
if __name__ == "__main__":
    # Use assim:
    analisar_rna("python-analise-dados/dados/bacteria.fasta", "bacteria.html")
    analisar_rna("python-analise-dados/dados/human.fasta", "human.html")
    # Ou com outro arquivo:
    # from funcao-compara-rna import analisar_rna
    #analisar_rna("dados/humano.fasta")  # Você chama manualmente
    

File: python-analise-dados/java/FuncaoComparaRna.java
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class FuncaoComparaRna {

    private static final char[] NUCLEOTIDE = {'A','T','C','G'};

    // Carrega sequência do FASTA removendo cabeçalho e quebras de linha
    public static String carregarSequencia(String caminhoFasta) throws IOException {
        Path p = Paths.get(caminhoFasta);
        List<String> linhas = Files.readAllLines(p);
        StringBuilder sb = new StringBuilder();
        for (String linha : linhas) {
            if (!linha.startsWith(">")) {
                sb.append(linha.trim());
            }
        }
        return sb.toString().toUpperCase();
    }

    // Cria dicionário (Map) com todos os dinucleotídeos possíveis inicializados em 0
    public static Map<String,Integer> criarDicionarioDinucleotideos() {
        Map<String,Integer> cont = new LinkedHashMap<>();
        for (char a : NUCLEOTIDE) {
            for (char b : NUCLEOTIDE) {
                cont.put("" + a + b, 0);
            }
        }
        return cont;
    }

    // Conta dinucleotídeos na sequência (ignora pares não presentes no dicionário)
    public static void contarDinucleotideos(String sequencia, Map<String,Integer> cont) {
        int len = sequencia.length();
        for (int i = 0; i < len - 1; i++) {
            String par = sequencia.substring(i, i + 2);
            if (cont.containsKey(par)) {
                cont.put(par, cont.get(par) + 1);
            }
        }
    }

    
    // Gera arquivo HTML simples com blocos coloridos representando as contagens
    public static void gerarHtml(Map<String,Integer> contagem, String arquivoSaida) throws IOException {
        int maxVal = contagem.values().stream().mapToInt(Integer::intValue).max().orElse(1);
        Path out = Paths.get(arquivoSaida);
        try (PrintWriter pw = new PrintWriter(Files.newBufferedWriter(out))) {
            pw.println("<html><body>");
            int i = 1;
            for (Map.Entry<String,Integer> e : contagem.entrySet()) {
                double transparencia = (maxVal > 0) ? ((double) e.getValue() / maxVal) : 0.0;
                // Força ponto decimal usando Locale.US
                String alpha = String.format(Locale.US, "%.6f", transparencia);
                

                pw.println("<div style='width:100px; border:1px solid #111; height:100px; float:left; "
                        + "background-color:rgba(0, 0, 255, " + alpha + ")'>"
                        + e.getKey() + "-" + e.getValue() + "</div>");
                
                if (i % 4 == 0) {
                    pw.println("<div style='clear:both'></div>");
                }
                i++;
            }
            pw.println("</body></html>");
        }
    }


    // Função principal que orquestra a análise
    public static void analisarRna(String caminhoFasta, String arquivoSaida) {
        try {
            System.out.println("Carregando sequência de " + caminhoFasta + "...");
            String seq = carregarSequencia(caminhoFasta);

            System.out.println("Criando dicionário de dinucleotídeos...");
            Map<String,Integer> cont = criarDicionarioDinucleotideos();
            System.out.println("Dicionário inicial: " + cont);

            System.out.println("Contando dinucleotídeos...");
            contarDinucleotideos(seq, cont);
            System.out.println("Contagem final: " + cont);

            System.out.println("Gerando arquivo HTML: " + arquivoSaida + "...");
            gerarHtml(cont, arquivoSaida);
            System.out.println("✅ Análise concluída com sucesso!");
        } catch (IOException ex) {
            System.err.println("Erro: " + ex.getMessage());
        }
    }


    public static void main(String[] args) {
        // Exemplo de uso: se passar argumentos, usa-os; senão usa caminhos padrão
        if (args.length >= 2) {
            analisarRna(args[0], args[1]);
        } else {
            analisarRna("python-analise-dados/dados/bacteria.fasta", "bacteria.html");
            analisarRna("python-analise-dados/dados/human.fasta", "human.html");
        }
    }
}